### PR TITLE
DashboardScene: Update query variable definition on query change

### LINKED ISSUE #81631: When modifying a variable (prometheus variable in my case) and navigating back to the variables list, the variable definition is not updated
I found one issue that I think **may not be** related to the scope of this PR. When modifying a variable (prometheus variable in my case) and navigating back to the variables list, the variable definition is not updated. See the attached video:

https://github.com/grafana/grafana/assets/2376619/04336738-c227-45b8-9530-10c987a4331c

Discovered here: https://github.com/grafana/grafana/pull/80847#pullrequestreview-1843801248

### PR DESCRIPTION
**Pull request by Sergej-Vlasov**
- [x] adjusted query variable definition on query change

**Which issue(s) does this PR fix?**:

Fixes #81631

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard-scene/settings/variables/editors/QueryVariableEditor.tsx [hunk 1]
```diff
@@ -50,7 +50,15 @@ export function QueryVariableEditor({ variable, onRunQuery }: QueryVariableEdito
...
-    variable.setState({ query });
+    let definition: string;
+    if (typeof query === 'string') {
+      definition = query;
+    } else if (query.hasOwnProperty('query') && typeof query.query === 'string') {
+      definition = query.query;
+    } else {
+      definition = '';
+    }
+    variable.setState({ query, definition });
```
Comment (Sergej-Vlasov) [inline public/app/features/dashboard-scene/settings/variables/editors/QueryVariableEditor.tsx:61]: @ivanortegaalba just updating to string was not enough, so I found additional logic in `public/app/features/dashboard-scene/settings/variables/editors/QueryVariableEditor.tsx` `onQueryChange` function and added it here as well

**Review by axelavargas — COMMENTED** [review]
Great work @Sergej-Vlasov 🙌🏾, the code looks good to me, we are missing adding some unit tests to the current [suite](https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/settings/variables/editors/QueryVariableEditor.test.tsx) 😄.

I also gave it a try locally and it works as expected.

**Comment by Sergej-Vlasov** [comment]
> Great work @Sergej-Vlasov 🙌🏾, the code looks good to me, we are missing adding some unit tests to the current [suite](https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/settings/variables/editors/QueryVariableEditor.test.tsx) 😄.
> 
> I also gave it a try locally and it works as expected.

@axelavargas I added a test but it only covers the basic scenario when `typeof query === 'string'`. I tried adjusting the mocks in `setup()` function but could not achieve an actual ui change to the prometheus datasource where query becomes an object. If this is enough then all good, but if not then would need to pair with someone on this

**Review by axelavargas — APPROVED** [review]
@Sergej-Vlasov great work again with this one 🏅

Regarding the test, lets keep this one with the simple test case :)

**Pull request merged**